### ISSUE TITLE
AWSS3TransferUtility upload is not working with new Amplify

### ISSUE THREAD

**Issue by sudhar99** [aws-amplify/aws-sdk-ios#2067]
My older implementation without amplify used to work fine in S3  with my bucket name and prefix folder.
I have used the same setup using amplify with custom UI and manual entries. The folder name (key) in the bucket doesn't have the amplify style of { "protected", private etc.,..} but I changed my authRole to have write/delete/permission at the bucket level ("/bucket-1/*).

**Environment(please complete the following information):**

SDK Version: 2.12.1
Dependency Manager: Cocoapods
Swift Version : 5.0

Here is my aws-configuration.json file.
{
    "UserAgent": "aws-amplify/cli",
    "Version": "0.1.0",
    "IdentityManager": {
        "Default": {}
    },
    "CredentialsProvider": {
        "CognitoIdentity": {
            "Default": {
                "PoolId": "us-west-2:aaaaaaa",
                "Region": "us-west-2"
            }
        }
    },
    "CognitoUserPool": {
        "Default": {
            "PoolId": "us-west-2_xyz",
            "AppClientId": "yyyyy",
            "AppClientSecret": "xxxxx",
            "Region": "us-west-2"
        }
    },
    "S3TransferUtility": {
        "Default": {
            "Bucket": "bucket-1",
            "Region": "us-west-2"
        }
    },
    "S3": {
        "Default": {
            "Bucket": "bucket-1",
            "Region": "us-west-2"
        }
    }
} 

My custom UI signinviewController uses the amplify for all authentication and works fine.
AWSMobileClient.default().signIn(...)

Here is my upload with 2 different approaches:
1. Approach 1: Using the AWSS3TransferUtility.default()
error	NSError	domain: "com.amazonaws.AWSCognitoIdentityErrorDomain" - code: 8	

 let transferUtility = AWSS3TransferUtility.default() and try to upload transferUtility.upload(...
or by registering with the following and let transferUtility = AWSS3TransferUtility.s3TransferUtility(forKey: UNIQUE_KEY_S3_TRANSFER) and upload....
func registerS3TransferUtility()
{
    let transferUtilityConfigurationWithRetry = AWSS3TransferUtilityConfiguration()
    transferUtilityConfigurationWithRetry.isAccelerateModeEnabled = false
    transferUtilityConfigurationWithRetry.retryLimit = 10
    transferUtilityConfigurationWithRetry.multiPartConcurrencyLimit = 5
    transferUtilityConfigurationWithRetry.timeoutIntervalForResource = 15*60 //15 minutes
    // Obtain the reference to the AWSCredentialsProvider and AWSConfiguration objects   
    AWSS3TransferUtility.register(with: AWSS3TransferUtility.default().configuration,
        transferUtilityConfiguration: transferUtilityConfigurationWithRetry,
        forKey: UNIQUE_KEY_S3_TRANSFER)
}

2. Approach 2: Use AWSMobileClient.default() with userstate.
 com.amazonaws.AWSS3TransferUtilityErrorDomain error 2 
AWSMobileClient.default().initialize { (userState, error) in
        guard error == nil else {
            print("Error initializing AWSMobileClient. Error: \(error!.localizedDescription)")
            return
        }
        if let userState = userState{
            registerS3TransferUtility()
        }
    }

    let transferUtility = AWSS3TransferUtility.s3TransferUtility(forKey: UNIQUE_KEY_S3_TRANSFER)

func registerS3TransferUtility()
{
    let transferUtilityConfigurationWithRetry = AWSS3TransferUtilityConfiguration()
    //transferUtilityConfigurationWithRetry.isAccelerateModeEnabled = false
    transferUtilityConfigurationWithRetry.retryLimit = 5
    //transferUtilityConfigurationWithRetry.multiPartConcurrencyLimit = 5
    transferUtilityConfigurationWithRetry.timeoutIntervalForResource = 15*60 //15 minutes
    // Obtain the reference to the AWSCredentialsProvider and AWSConfiguration objects
    let configuration = AWSServiceConfiguration(region: AWSRegionType.USWest2, credentialsProvider: AWSMobileClient.default())
    AWSS3TransferUtility.register(with: configuration!,
        transferUtilityConfiguration: transferUtilityConfigurationWithRetry,
        forKey: UNIQUE_KEY_S3_TRANSFER)
}

- Bottom line, I don't what is the exact sequence to register the AWSS3TransferUtility for successful upload. The documentation is very vague and NOT helping at all.

**Comment by lawmicha** [comment]
Hi @sudhar99, 

From your code snippet, it looks like you are trying to set up TransferUtility with the credentials provider. The second approach looks good with AWSMobileClient

You can initialize AWSMobileClient using `AWSMobileClient.default().initialize`
```
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

    AWSMobileClient.default().initialize { (userState, error) in
        guard error == nil else {
            print("Error initializing AWSMobileClient. Error: \(error!.localizedDescription)")
            return
        }
        print("AWSMobileClient initialized.")
        // registerS3TransferUtility()
    }
    return true
}
```

2. You should be able to use the default transfer utility instance, it internally uses the default credentials provider.
```
AWSS3TransferUtility.default()
```

3. To configure a different instance of AWSS3TransferUtility to override default transfer utility configuration. You're second approach looks right:
```
func registerS3TransferUtility() {
        let transferUtilityConfigurationWithRetry = AWSS3TransferUtilityConfiguration()
        transferUtilityConfigurationWithRetry.retryLimit = 5
        transferUtilityConfigurationWithRetry.timeoutIntervalForResource = 15*60 //15 minutes
        transferUtilityConfigurationWithRetry.bucket = "bucket"
        // Obtain the reference to the AWSCredentialsProvider and AWSConfiguration objects
        let configuration = AWSServiceConfiguration(region: AWSRegionType.USWest2,
                                                    credentialsProvider: AWSMobileClient.default())
        AWSS3TransferUtility.register(with: configuration!,
                                      transferUtilityConfiguration: transferUtilityConfigurationWithRetry,
                                      forKey: "identifier")
    }
```
I modified it slightly to include the bucket. Otherwise you'll have to pass it in everytime when making a call.

4. Then you can use the instance by referencing it using the key that you provided when registering it:
`AWSS3TransferUtility.s3TransferUtility(forKey: "identifier").upload|download`

Regarding the auth role, `amplify storage` guides you to create the S3 bucket with your desired access levels. I can see many policies were created. ie.
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::<bucket>/<accessLevel>/${cognito-identity.amazonaws.com:sub}/*"
            ],
            "Effect": "Allow"
        }
    ]
}
```
If you want to make it completely public, can you clarify that the other policies are deleted and the "public" policy's resource would is updated to `arn:aws:s3:::<bucket>/*` ?

Hope this help, my suggestion would be to allow Amplify-CLI to help you provision the S3 with the correct policies, make sure your code is working that you have connectivity to S3, and then you can play around with the policies. Can you try again and let us know the method and associated code snippet you are making that is failing and printing out the error message? It's not super clear where it's failing for you.

**Comment by lawmicha** [comment]
there is also a sample here https://github.com/awslabs/aws-sdk-ios-samples/tree/master/S3TransferUtility-Sample/Swift

**Comment by XingZhaoDev** [comment]
Hi, I have Amplify iOS SDK installed in my project, is it still available to use AWSS3TransferUtility? Thanks.
